README.md
<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-40README-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

# cl-brewer - Homebrew formula builder for Common Lisp applications.

<a id="cl-brewer-asdf-system-details"></a>

## CL-BREWER ASDF System Details

* Description: Homebrew formula builder for Common Lisp applications.

* Licence: Public Domain

* Author: Dmitry Petrov <dpetroff@gmail.com>

* Maintainer: Alexander Artemenko <svetlyak.40wt@gmail.com>

* Homepage: [https://40ants.com/cl-brewer/][1887]

* Bug tracker: [https://github.com/40ants/cl-brewer/issues][557c]

* Source control: [GIT][df7d]

* Depends on: [alexandria][8236], [cl-plus-ssl-osx-fix][0ce0], [command-line-arguments][bdb0], [deploy][e073], [ironclad][90b9], [quicklisp][9c78], [str][ef7f], [trivial-download][923b]

[![](https://github-actions.40ants.com/40ants/cl-brewer/matrix.svg?only=ci.run-tests)][a54e]

![](http://quickdocs.org/badge/cl-brewer.svg)

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40ABOUT-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## About

Currently there is now easy way to distribute common lisp applications.
One promising way is to use [roswell][795a] project
however it might be to complicated for the users how just want to install
an application and are not really interested in having one more package manager
for that.

In Mac `OS` X world the most popular solution is to use brew package manager and all
we need is to be able to generate formula that will handle installation proccess.
Since homebrew guidelines are not really fond of using third-party managers to
get dependencies, we need to generate list manually and feed it to brew.

cl-brewer uses `SBCL` and targets command line applications written in Common Lisp.
`GUI` applications haven't been tested and might require additional changes.

This application is based on the awesome [quicklisp-homebrew-roundup][f9c6]
but has a purpose to make a process simplier by:

* Allowing to generate file for any system available not just quicklisp package (dependencies should be on quicklisp though)

* Allowing to generate a complete formula, not just dependencies

* Having a cli interface that can be used in `CI` services to generate formulas automatically.

At the moment several assumptions were made:

* We can use [Buildapp][ebe7] or [Deploy][eaea] make executable.

* Formula is generated for the system available for `ASDF`.

* All dependencies should be available via quicklisp.

* System source code should live on github all releases should be tagged with vX.Y.Z scheme.

* System should have description, version, homepage fields defined. Project at GitHub should have a tag corresponding to the
  current system version. GitHub build a tar.gz archive with sources of the tagged revision and this way a formula,
  created by cl-brewer, will be able to fetch sources, corresponding to the version.

* [Buildapp][ebe7] calls `main` function. By default `main` function is searched in the
  package with the same name as system name but can be overridden with option.

* When using [Deploy][eaea], you don't have to define an entry point. Also, Deploy is able to
  create formulas for applications which build some dynamic libraries.

Here is an [example formula][cccc]

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40INSTALLATION-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## Installation

```
brew tap 40ants/soft
brew install cl-brewer
```
Or you can install it using Roswell:

```
# install roswell and sbcl before
$ ros install 40ants/cl-brewer
```
If you want to install it to use from the `REPL`, then you can install this library from Quicklisp,
but you want to receive updates quickly, then install it from Ultralisp.org:

```
(ql-dist:install-dist "http://dist.ultralisp.org/"
                      :prompt nil)
(ql:quickload :cl-brewer)
```
<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40USAGE-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## Usage

Just run:

```
cl-brewer <your-system-name>
```
This will emit `<your-system-name>.rb` file in current folder.

To run cl-brewer under Qlot, you'll need set `SBCL_HOME` variable.
Otherwise `sb-sys::*sbcl-homedir-pathname*` variable will be reinitialized
and this can cause some incompatibilities if Qlot and cl-brewer were built
with different `SBCL` implementations.

Thus, in this case call cl-brewer like this:

`
qlot exec bash -c "SBCL_HOME='' cl-brewer <your-system-name>"
` 

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40BUILDING-CL-BREWER-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## Building a formula for cl-brewer

To kickstart a cl-brewer and to create a formula for itself, load it in the `REPL` and do like that:

```
CL-USER> (cl-brewer:create-formula :cl-brewer)
#<BUILDAPP-FORMULA "cl-brewer" depends on 23 systems>

CL-USER> (cl-brewer:save-formula * "cl-brewer"
                                 :preload (list "quicklisp-starter"))
Downloading "https://github.com/40ants/cl-brewer/archive/v0.5.5.tar.gz" (Unknown size)
NIL
```
However, in most cases you can just install cl-brewer from the Homebrew. In this case,
you can update `cl-brewer's` formula with this command:

```
qlot exec cl-brewer 
          --preload quicklisp-starter 
          cl-brewer
```
<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40LOCAL-INSTALL-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## Installing From Local Formula

How to install cl-brewer (or any other project) from a local formula?

Replace url line in a formula:

```
url "https://github.com/40ants/cl-brewer/archive/v0.5.6.tar.gz"
```
with two lines like this:

```
url File.dirname(__FILE__), :using => :git
version "0.5.6-rc1"
```
Next, do this in the shell:

```
HOMEBREW_NO_AUTO_UPDATE=1 brew install --debug --verbose ./*.rb
```
it should build and install `cl-brewer`.

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40CONTRIBUTE-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## Contribute

If you are interested in using this project but your application has different requirements,
please open an issue or make a pull request. Contributions are welcome!

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40LICENSE-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## License

All code is public domain except parts that were taken from [quicklisp-homebrew-roundup][f9c6] which is under `MIT` License.

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40API-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## API

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40CL-BREWER-3FPACKAGE-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

### CL-BREWER

<a id="x-28-23A-28-289-29-20BASE-CHAR-20-2E-20-22CL-BREWER-22-29-20PACKAGE-29"></a>

#### [package](67c2) `cl-brewer`

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-7C-40CL-BREWER-3FClasses-SECTION-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

#### Classes

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40CL-BREWER-24FORMULA-3FCLASS-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

##### FORMULA

<a id="x-28CL-BREWER-3AFORMULA-20CLASS-29"></a>

###### [class](ba10) `cl-brewer:formula` ()

Base class for Homebrew formula definition.

**Readers**

<a id="x-28CL-BREWER-2FFORMULA-3AINCLUDED-SYSTEMS-20-2840ANTS-DOC-2FLOCATIVES-3AREADER-20CL-BREWER-3AFORMULA-29-29"></a>

###### [reader](f60a) `cl-brewer/formula:included-systems` (formula) (:included-systems)

<a id="x-28CL-BREWER-2FFORMULA-3AMISSING-SYSTEMS-20-2840ANTS-DOC-2FLOCATIVES-3AREADER-20CL-BREWER-3AFORMULA-29-29"></a>

###### [reader](b64d) `cl-brewer/formula:missing-systems` (formula) (:missing-systems)

<a id="x-28CL-BREWER-2FFORMULA-3AROOT-SYSTEM-20-2840ANTS-DOC-2FLOCATIVES-3AREADER-20CL-BREWER-3AFORMULA-29-29"></a>

###### [reader](8494) `cl-brewer/formula:root-system` (formula) (:root-system)

**Accessors**

<a id="x-28CL-BREWER-2FFORMULA-3AINCLUDED-SYSTEMS-20-2840ANTS-DOC-2FLOCATIVES-3AACCESSOR-20CL-BREWER-3AFORMULA-29-29"></a>

###### [accessor](f60a) `cl-brewer/formula:included-systems` (formula) (:included-systems)

<a id="x-28CL-BREWER-2FFORMULA-3AMISSING-SYSTEMS-20-2840ANTS-DOC-2FLOCATIVES-3AACCESSOR-20CL-BREWER-3AFORMULA-29-29"></a>

###### [accessor](b64d) `cl-brewer/formula:missing-systems` (formula) (:missing-systems)

<a id="x-28CL-BREWER-2FFORMULA-3AROOT-SYSTEM-20-2840ANTS-DOC-2FLOCATIVES-3AACCESSOR-20CL-BREWER-3AFORMULA-29-29"></a>

###### [accessor](8494) `cl-brewer/formula:root-system` (formula) (:root-system)

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-7C-40CL-BREWER-3FGenerics-SECTION-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

#### Generics

<a id="x-28CL-BREWER-3ACREATE-FORMULA-20GENERIC-FUNCTION-29"></a>

##### [generic-function](0323) `cl-brewer:create-formula` system

Create <formula> object based on asdf:system with a list of all dependencies

<a id="x-28CL-BREWER-3AGET-IMPLICIT-DEPENDENCIES-20GENERIC-FUNCTION-29"></a>

##### [generic-function](775c) `cl-brewer:get-implicit-dependencies` system-name

Some systems, like cl-unicode have implicit dependencies in their asdf methods:
[https://github.com/edicl/cl-unicode/blob/8073fc5634c9d4802888ac03abf11dfe383e16fa/cl-unicode.asd#L67-L70][44a6]
use this method to provide information about such dependencies.

System name is a keyword and method should return a one keyword or a list of keywords with names of systems.
Each returned system should be possible to find with ql-dist:find-system.

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-7C-40CL-BREWER-3FFunctions-SECTION-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

#### Functions

<a id="x-28CL-BREWER-3ASAVE-FORMULA-20FUNCTION-29"></a>

##### [function](b50c) `cl-brewer:save-formula` formula name &key entry-point preload

Saves Homebrew formula definition into the file with given `NAME`.

If `ENTRY-POINT` argument was given, then it might be used as entry-point,
but some formula classes like [`cl-brewer/deploy/formula:deploy-formula`][de5a]
might ignore this argument.

`PRELOAD` argument if given, should be a list of strings with
`ASDF` system names to be preloaded before cl-brewer will build a binary.

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40CL-BREWER-2FBUILDAPP-2FFORMULA-3FPACKAGE-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

### CL-BREWER/BUILDAPP/FORMULA

<a id="x-28-23A-28-2826-29-20BASE-CHAR-20-2E-20-22CL-BREWER-2FBUILDAPP-2FFORMULA-22-29-20PACKAGE-29"></a>

#### [package](b375) `cl-brewer/buildapp/formula`

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-7C-40CL-BREWER-2FBUILDAPP-2FFORMULA-3FClasses-SECTION-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

#### Classes

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40CL-BREWER-2FBUILDAPP-2FFORMULA-24BUILDAPP-FORMULA-3FCLASS-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

##### BUILDAPP-FORMULA

<a id="x-28CL-BREWER-2FBUILDAPP-2FFORMULA-3ABUILDAPP-FORMULA-20CLASS-29"></a>

###### [class](7edb) `cl-brewer/buildapp/formula:buildapp-formula` (formula)

This formula class uses [Buildapp][ebe7] to build a binary.

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40CL-BREWER-2FDEPLOY-2FFORMULA-3FPACKAGE-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

### CL-BREWER/DEPLOY/FORMULA

<a id="x-28-23A-28-2824-29-20BASE-CHAR-20-2E-20-22CL-BREWER-2FDEPLOY-2FFORMULA-22-29-20PACKAGE-29"></a>

#### [package](3dfc) `cl-brewer/deploy/formula`

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-7C-40CL-BREWER-2FDEPLOY-2FFORMULA-3FClasses-SECTION-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

#### Classes

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40CL-BREWER-2FDEPLOY-2FFORMULA-24DEPLOY-FORMULA-3FCLASS-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

##### DEPLOY-FORMULA

<a id="x-28CL-BREWER-2FDEPLOY-2FFORMULA-3ADEPLOY-FORMULA-20CLASS-29"></a>

###### [class](b11c) `cl-brewer/deploy/formula:deploy-formula` (formula)

This formula class uses [Deploy][eaea] to build a binary.

The core difference from [`cl-brewer/buildapp/formula:buildapp-formula`][854c] is that
this type of formula also builds and distributes all necessary dynamic libraries.

<a id="x-28CL-BREWER-DOCS-2FINDEX-3A-3A-40CL-BREWER-2FFORMULA-3FPACKAGE-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

### CL-BREWER/FORMULA


Changelog.md
<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-40CHANGELOG-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

# ChangeLog

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E11-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.11.0 (2023-07-26)

* Now `CL_SOURCE_REGISTRY` includes current dir non-recursively and all dependencies
  are placed in the `_brew_resources` folder instead of `lib`.

* Also, all warnings are muffled on compilation.

* And `--formula-name` option was added. This way you'll be able to override filename
  of the formula.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E10-2E5-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.10.5 (2023-07-25)

* Separated system with hooks into it's own `ASDF` system, solved a fiew other build errors.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E10-2E4-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.10.4 (2023-07-24)

* Cleanup and use newer 40ants-asdf-system for proper `--version` handling.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E10-2E3-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.10.3 (2023-07-23)

* Fixed arguments parsing.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E10-2E2-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.10.2 (2023-07-23)

* Fixed how cl-brewer's version number is stored in the binary.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E10-2E1-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.10.1 (2023-07-23)

* Fixed build commands escaping for Deploy formula.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E10-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.10.0 (2023-07-23)

* Library was refactored to use package-inferred system style. Also a new style documentation was added.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E9-2E1-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.9.1 (2023-07-14)

* Fixed an error happened in case if program does not have any dynamic libs to install into the Cellar.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E9-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.9.0 (2023-07-14)

Now when using Deploy, dynamic libraries copied into the libexec
folder inside Cellar and not symlinked into the /ope/homebrew/bin folder.

Also, dynamic libraries provided by other Homebrew formulas are not
copied into the Cellar, instead these formulas are listed as a dependency
of the formula, created by cl-homebrew.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E8-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.8.0 (2019-09-12)

* Now cl-brewer suppress debug output of the Deploy's startup process.

* Added a `--version` command line option.

* Fixed creation of dist, local-projects and tmp in the current directory.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E7-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.7.0 (2019-09-11)

* Added a generic-function [`cl-brewer:get-implicit-dependencies`][dda5] which can be used to
specify additional dependencies for systems when asdf is not possible to figure them out.

It can be used like that:

```
(defmethod cl-brewer:get-implicit-dependencies ((system-name (eql :cl-unicode)))
    :flexi-streams)
```
It can return a single keyword or a list of keywords.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E6-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.6.0 (2019-07-30)

* Now cl-brewer can be installed from the Homebrew tap as a binary.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E5-2E6-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.5.6 (2019-07-22)

* Fixed arguments list in buildapp's entrypoint.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E5-2E5-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.5.5 (2019-07-22)

* Fixed a bug in preloading code on buildapp.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E5-2E4-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.5.4 (2019-07-22)

* Added support for `--preload` option.
It allows to embedd Quicklisp client when building a binary for a homebrew.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E5-2E3-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.5.3 (2019-07-22)

* More diagnostics to understand why does not work a binary built by homebrew.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E5-2E2-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.5.2 (2019-07-22)

* Added an entry point for buildapp.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E5-2E1-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.5.1 (2019-07-21)

* Kickstarting quicklisp client when building cl-brewer.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E5-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.5.0 (2019-07-21)

* Now cl-brewer can work with package-inferred asdf systems.

* Previously, some archives were included more than once because
  they contains multiple systems. Now this is fixed.

* Added option `--compress-core`. It reduces size of a simple
  "hello world" from `52M` to `13M`.

* Function `cl-brewer::create-formula` was made external and now
  can accept system name as a symbol.

* Function `cl-brewer::save-formula` was made external.

* `ASDF` option `defsystem-depends-on` was supported.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E4-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.4.0 (2016-12-17)

* Provide a way to skip some systems if not found by quicklisp. Right now it's only sb-introspect

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E3-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.3.0 (2016-11-10)

* A bit of doc strings.

* Cli interface.

* Code was split into multiple files.

<a id="x-28CL-BREWER-DOCS-2FCHANGELOG-3A-3A-7C0-2E2-2E0-7C-2040ANTS-DOC-2FLOCATIVES-3ASECTION-29"></a>

## 0.2.0 (2016-11-04)

Initial working functionality.


[dda5]: https://40ants.com/cl-brewer/#x-28CL-BREWER-3AGET-IMPLICIT-DEPENDENCIES-20GENERIC-FUNCTION-29

* * *
###### [generated by [40ANTS-DOC](https://40ants.com/doc/)]


cl-brewer.rb
class ClBrewer < Formula
  desc "Homebrew formula builder for Common Lisp applications."
  homepage "https://40ants.com/cl-brewer/"
  url "https://github.com/40ants/cl-brewer/archive/v0.11.0.tar.gz"
  sha256 "a465bd2a5737e92a505bd01461e2e50b74a71062376557170acb5de6de3ecbb9"
  head "https://github.com/40ants/cl-brewer"

  depends_on "sbcl" => :build
  # required by: CL+SSL/CONFIG::LIBCRYPTO and CL+SSL/CONFIG::LIBSSL
  depends_on "openssl@3"
  # required by: DEPLOY::COMPRESSION-LIB
  depends_on "zstd"

  resource "40ants-40ants-asdf-system" do
    url "http://dist.ultralisp.org/archive/1825/40ants-40ants-asdf-system-20230724171226.tgz"
    sha256 "6a1f4785fb233bdf38fe4cbcb6e8181b9a3e6b3e6790eaa16e6f92dbfd4e3ab0"
  end

  resource "40ants-cl-brewer" do
    url "http://dist.ultralisp.org/archive/1977/40ants-cl-brewer-20230725165337.tgz"
    sha256 "10b4d88c0d79a920230b012a8b3e12ef544534003d507e70da753ea44893bff9"
  end

  resource "40ants-cl-plus-ssl-osx-fix" do
    url "http://dist.ultralisp.org/archive/1962/40ants-cl-plus-ssl-osx-fix-20230618185126.tgz"
    sha256 "f26673e427545bfc7affae80d14d74a78c6b5f17b1a68d4a38c6223e894f92bb"
  end

  resource "alexandria" do
    url "http://beta.quicklisp.org/archive/alexandria/2023-06-18/alexandria-20230618-git.tgz"
    sha256 "d5fbb349204082b61a7a6c58aed1d1606e524bfa38ecd0ec6ebed7f658061819"
  end

  resource "cffi" do
    url "http://beta.quicklisp.org/archive/cffi/2023-06-18/cffi-20230618-git.tgz"
    sha256 "223f9fa6f1c50df1a8d5b7db0d02c520a4454965ed1a996cf5a6d20f447058c3"
  end

  resource "cl+ssl" do
    url "http://beta.quicklisp.org/archive/cl+ssl/2023-06-18/cl+ssl-20230618-git.tgz"
    sha256 "c5653c3d96b169003a386628cf0c92aaf256ba27377f569fe45e4e1012c80343"
  end

  resource "cl-babel-babel" do
    url "http://dist.ultralisp.org/archive/1237/cl-babel-babel-20230131043757.tgz"
    sha256 "cd3c1154be061834514a41ccc826735edfdf376ea9124e338f54209861a8a1b1"
  end

  resource "cl-base64" do
    url "http://beta.quicklisp.org/archive/cl-base64/2020-10-16/cl-base64-20201016-git.tgz"
    sha256 "3ff50faf5ddccd409f8954eb70c2d4e76329cc916f070de95f79c7ecf6d3a2f1"
  end

  resource "cl-change-case" do
    url "http://beta.quicklisp.org/archive/cl-change-case/2021-04-11/cl-change-case-20210411-git.tgz"
    sha256 "b2920b37fe4eae7864c0c2674405cb84ed951891462edcb8d536c003d2324293"
  end

  resource "command-line-arguments" do
    url "http://beta.quicklisp.org/archive/command-line-arguments/2021-08-07/command-line-arguments-20210807-git.tgz"
    sha256 "939b3966e2887dd0b81bd1c3d051c42bce78c10fa32661263c9aa6c355fbf9bd"
  end

  resource "edicl-chunga" do
    url "http://dist.ultralisp.org/archive/1244/edicl-chunga-20220912164619.tgz"
    sha256 "a34e2b85c76b3633f4f5dba26548a116d2b64f15db7c7e31ba04f913baf251c8"
  end

  resource "edicl-cl-ppcre" do
    url "http://dist.ultralisp.org/archive/1239/edicl-cl-ppcre-20230614075200.tgz"
    sha256 "fcf1d505007aa0fce3d09752c5fd1459efbaae43fb80f06980e591c7ec393918"
  end

  resource "edicl-cl-unicode" do
    url "http://dist.ultralisp.org/archive/1241/edicl-cl-unicode-20210222215228.tgz"
    sha256 "8bbf6b8ff8d51caa9e3c329509345f59636cbd61e8c896112c792bbcbd3a0751"
  end

  resource "edicl-drakma" do
    url "http://dist.ultralisp.org/archive/839/edicl-drakma-20230221010508.tgz"
    sha256 "d741be9ce6ccb46013d73f01b873fbd8576997c6cb97048695e8e181f97c6082"
  end

  resource "edicl-flexi-streams" do
    url "http://dist.ultralisp.org/archive/1242/edicl-flexi-streams-20220112225522.tgz"
    sha256 "bfbb31674c02abdea0ed8135e1894a2f743df4b03e8bab7db1506b90270d95f9"
  end

  resource "eudoxia0-trivial-download" do
    url "http://dist.ultralisp.org/archive/202/eudoxia0-trivial-download-20230124040434.tgz"
    sha256 "03e7abf5b523a127537abdca9fb1e6a379d2d69c54a2febdfa2dd8ffe67a5eee"
  end

  resource "lmj-global-vars" do
    url "http://dist.ultralisp.org/ultralisp/archive/l/lmj-global-vars-20190319075150.tgz"
    sha256 "b8a52b9ef2152f4087c1bf1ea9ea29f137419aa81d6ce4e250af2c34b169168d"
  end

  resource "puri" do
    url "http://beta.quicklisp.org/archive/puri/2020-10-16/puri-20201016-git.tgz"
    sha256 "fd3bd9acd2438eb89fcd0c8651b6f6b5d92f337872a8e59326a680666e7c079a"
  end

  resource "quicklisp-quicklisp-client" do
    url "http://dist.ultralisp.org/archive/1254/quicklisp-quicklisp-client-20210216010650.tgz"
    sha256 "d8e39e878da43e54ea69194c075da371e57c7c76f01000d541dc43e3c3ee6721"
  end

  resource "sharplispers-chipz" do
    url "http://dist.ultralisp.org/archive/1771/sharplispers-chipz-20230418140035.tgz"
    sha256 "62f582c1904782dff1e23995b80cf5ad7faa4d0408a7a096cb46ade63d2fe991"
  end

  resource "sharplispers-ironclad" do
    url "http://dist.ultralisp.org/archive/655/sharplispers-ironclad-20230710201532.tgz"
    sha256 "769001a6509105daa01df15427ca28147c8c59249a2a69ff0f94ca5a98335442"
  end

  resource "sharplispers-split-sequence" do
    url "http://dist.ultralisp.org/archive/273/sharplispers-split-sequence-20211208061629.tgz"
    sha256 "98c27530444a65fcdc71f2dc539a252efa977f1a2ebfd6a0453ddfa572ca1d83"
  end

  resource "Shinmera-deploy" do
    url "http://dist.ultralisp.org/archive/858/Shinmera-deploy-20230710201211.tgz"
    sha256 "1ad782e17010c250f69965b514add3cc5fffa3fddba3066c9d062007905c462b"
  end

  resource "Shinmera-documentation-utils" do
    url "http://dist.ultralisp.org/archive/843/Shinmera-documentation-utils-20230711001755.tgz"
    sha256 "e8d287de786216763a2fafbc7d18da539b7a124497b8529f066c33f0148f2182"
  end

  resource "Shinmera-trivial-indent" do
    url "http://dist.ultralisp.org/archive/193/Shinmera-trivial-indent-20230710204252.tgz"
    sha256 "7b4c8b6db7250fb2794a79967fcb5c0410146b206233e43227b708d92d3ec944"
  end

  resource "sionescu-bordeaux-threads" do
    url "http://dist.ultralisp.org/archive/1238/sionescu-bordeaux-threads-20230604143150.tgz"
    sha256 "388625b5f352f5099bffa548569582c6eeb751e0293b4e861c3e534643d43587"
  end

  resource "trivial-features-trivial-features" do
    url "http://dist.ultralisp.org/archive/197/trivial-features-trivial-features-20230614074348.tgz"
    sha256 "6dab2a6ee703a8d444d3e0438afc838138a8943c4ae1640cc5d710dabba35b34"
  end

  resource "trivial-garbage-trivial-garbage" do
    url "http://dist.ultralisp.org/archive/195/trivial-garbage-trivial-garbage-20230621104435.tgz"
    sha256 "debc94633569e61b4f12fbee52b127b4234385795e0dd2dde9099925a5ab66cd"
  end

  resource "trivial-gray-streams-trivial-gray-streams" do
    url "http://dist.ultralisp.org/archive/194/trivial-gray-streams-trivial-gray-streams-20230630171731.tgz"
    sha256 "52883684b77d874de095dd56752dda1226a54030b5ede28a3ae71de0ee93e284"
  end

  resource "uiop" do
    url "http://beta.quicklisp.org/archive/uiop/2023-06-18/uiop-3.3.6.tgz"
    sha256 "302acb92b985b4b44a2ae2bdcc0d385084138c17acaf2cdc7ed2dc155172ec70"
  end

  resource "usocket" do
    url "http://beta.quicklisp.org/archive/usocket/2023-06-18/usocket-0.8.6.tgz"
    sha256 "33b94a7d3b2258bf6f06fbc52560dfbc3d2d8dd87ef407bfb18f10c40b17eb6e"
  end

  resource "vindarel-cl-str" do
    url "http://dist.ultralisp.org/archive/36/vindarel-cl-str-20230710200441.tgz"
    sha256 "5619486c2b5184d6b1ca99c452ee63b393ddca6d1b007c373c262fb7e62408f7"
  end

  def install
    resources.each do |resource|
      resource.stage buildpath/"_brew_resources"/resource.name
    end

    ENV["LIBEXEC_PATH"] = "#{libexec}/"
    ENV["CL_SOURCE_REGISTRY"] = "#{buildpath}/:#{buildpath}/_brew_resources//"
    ENV["ASDF_OUTPUT_TRANSLATIONS"] = "/:/"

    system "sbcl", "--eval", "(require :asdf)", "--eval", "(push :deploy-console *features*)", "--eval", "(asdf:load-system :cl-brewer-deploy-hooks)", "--eval", "(HANDLER-BIND ((ERROR
                (LAMBDA (E)
                  (UIOP/IMAGE:PRINT-BACKTRACE :CONDITION E)
                  (UIOP/IMAGE:QUIT 1)))
               (WARNING #'MUFFLE-WARNING))
  (ASDF/OPERATE:LOAD-SYSTEM \"quicklisp-starter\"))", "--eval", "(HANDLER-BIND ((ERROR
                (LAMBDA (E)
                  (UIOP/IMAGE:PRINT-BACKTRACE :CONDITION E)
                  (UIOP/IMAGE:QUIT 1)))
               (WARNING #'MUFFLE-WARNING))
  (ASDF/OPERATE:LOAD-SYSTEM \"cl-plus-ssl-osx-fix\"))", "--eval", "(HANDLER-BIND ((ERROR
                (LAMBDA (E)
                  (UIOP/IMAGE:PRINT-BACKTRACE :CONDITION E)
                  (UIOP/IMAGE:QUIT 1)))
               (WARNING #'MUFFLE-WARNING))
  (ASDF/OPERATE:MAKE \"cl-brewer\"))"

    system "bash", "-c", "mkdir dyn-libs && find bin/ -name '*.dylib' -exec mv '{}' dyn-libs/ \\;"

    bin.install Dir["bin/*"]
    libexec.install Dir["dyn-libs/*"]
  end
end
